File: frontend/src/lib/api/orchestrator.ts
import http from "./client";
import type { OrchestratorResponse, TriageEventPayload } from "./types";

export interface HandleEventRequest {
  event: TriageEventPayload;
  primary_asset_id?: string;
  promote_threats?: boolean;
  run_threat_intel?: boolean;
  run_investigation?: boolean;
  run_response?: boolean;
  run_forensics?: boolean;
  run_compliance?: boolean;
  investigation_lookback_hours?: number;
}

// Run an event through the full agent pipeline.
// The full 6-stage run can take 60-90s (the LLM tier is rate-limited per
// minute), well past the client's default 60s timeout, so extend it here to
// avoid a false "timeout" failure while the backend is still working.
const PIPELINE_TIMEOUT_MS = 180_000;

export async function handleEvent(
  request: HandleEventRequest
): Promise<OrchestratorResponse> {
  const { data } = await http.post<OrchestratorResponse>(
    "/orchestrator/handle-event",
    request,
    { timeout: PIPELINE_TIMEOUT_MS }
  );
  return data;
}

File: backend/app/api/v1/endpoints/orchestrator.py
from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, status
from pydantic import BaseModel, Field

from app.agents.triage import TriageInput
from app.api.deps import CurrentUser, require_analyst
from app.core.logging import get_logger
from app.services.orchestrator_service import run_full_pipeline

router = APIRouter()
log = get_logger(__name__)


# Request / response

class OrchestratorRequest(BaseModel):
    event: TriageInput
    primary_asset_id: str | None = None
    promote_threats: bool = Field(default=True)
    run_threat_intel: bool = Field(default=True)
    run_investigation: bool = Field(default=True)
    run_response: bool = Field(default=True)
    run_forensics: bool = Field(default=True)
    run_compliance: bool = Field(default=True)
    investigation_lookback_hours: int = Field(default=168, ge=1, le=720)


class OrchestratorResponse(BaseModel):
    stages: dict[str, Any]
    summary: dict[str, Any]


# POST /orchestrator/handle-event

@router.post(
    "/handle-event",
    response_model=OrchestratorResponse,
    status_code=status.HTTP_200_OK,
    summary="Run an event through the full agent pipeline",
)
async def handle_event(
    request: OrchestratorRequest,
    user: CurrentUser = Depends(require_analyst),
) -> OrchestratorResponse:
    """Run an event through the full agent pipeline.

    Order is triage, threat intel, investigation, response, forensics, compliance.
    """
    result = run_full_pipeline(
        organization_id=user.organization_id,
        event=request.event,
        primary_asset_id=request.primary_asset_id,
        promote_threats=request.promote_threats,
        run_threat_intel=request.run_threat_intel,
        run_investigation=request.run_investigation,
        run_response=request.run_response,
        run_forensics=request.run_forensics,
        run_compliance=request.run_compliance,
        investigation_lookback_hours=request.investigation_lookback_hours,
    )
    return OrchestratorResponse(**result)